rotkehlchen/accounting/cost_basis/base.py

```diff
@@ -136,27 +151,85 @@ def consume_result(self, used_amount: FVal) -> None:
         # Written on 06.06.2022 and can be removed after a couple of months if everything goes well
-        assert ZERO <= used_amount <= self._acquisitions[0].remaining_amount, \
-            f'Used amount must be in the interval [0, {self._acquisitions[0].remaining_amount}] but it was {used_amount}'  # noqa: E501
+        assert ZERO <= used_amount <= self._acquisitions[0][1].remaining_amount, \
+            f'Used amount must be in the interval [0, {self._acquisitions[0][1].remaining_amount}] but it was {used_amount}'  # noqa: E501
 
-        self._acquisitions[0].remaining_amount -= used_amount
-        if self._acquisitions[0].remaining_amount == ZERO:
-            self._acquisitions.popleft()
+        self._acquisitions[0][1].remaining_amount -= used_amount
+        if self._acquisitions[0][1].remaining_amount == ZERO:
+            heapq.heappop(self._acquisitions)
 
     def __len__(self) -> int:
         return len(self._acquisitions)
 
 
 class FIFOAcquisitionsOrder(BaseAcquisitionsOrder):
     """Accounting in FIFO (first-in-first-out) order"""
+    _count: int
+
+    def __init__(self) -> None:
+        self._count = 0
+        super().__init__()
+
     def add_acquisition(self, acquisition: AssetAcquisitionEvent) -> None:
-        self._acquisitions.append(acquisition)
+        heapq.heappush(self._acquisitions, (self._count, acquisition))
+        self._count += 1
 
 
 class LIFOAcquisitionsOrder(BaseAcquisitionsOrder):
     """Accounting in LIFO (last-in-first-out) order"""
+    _count: int
+
+    def __init__(self) -> None:
+        self._count = 0
+        super().__init__()
+
+    def add_acquisition(self, acquisition: AssetAcquisitionEvent) -> None:
+        heapq.heappush(self._acquisitions, (-self._count, acquisition))
+        self._count += 1
+
+
+class HIFOAcquisitionsOrder(BaseAcquisitionsOrder):
+    """Accounting in HIFO (highest-in-first-out) order"""
+    def add_acquisition(self, acquisition: AssetAcquisitionEvent) -> None:
+        heapq.heappush(self._acquisitions, (-acquisition.amount, acquisition))
+
+
+class AverageCostBaseAcquisitionOrder(BaseAcquisitionsOrder):
+    """Accounting in Average Cost Base(ACB) method"""
+
+    # the current cost base of last event(buy/sell) that occurred
+    current_cost_base: FVal
+    # a variable that increments whenever .processing_iterator is called
+    _last_acquisition_index: int
+    # keeps track of the amount of the asset remaining after evert acquisition or spend
+    remaining_amount: FVal
+
+    def __init__(self) -> None:
+        self.current_cost_base = ZERO
+        self._last_acquisition_index = -1
+        self.remaining_amount = ZERO
+        super().__init__()
+
     def add_acquisition(self, acquisition: AssetAcquisitionEvent) -> None:
-        self._acquisitions.appendleft(acquisition)
+        heapq.heappush(self._acquisitions, (Timestamp(time.time_ns()), acquisition))
```

Comment:
Hm:
1. Why not to use counter instead of current time?
2. Why is it needed at all? Why to store acquisitions for average cost basis?